### PR TITLE
WIP fix: merge api challenge subscription close into api challenge subscr…

### LINKED ISSUE #530: get rid of the "close" call and autoclose on validate

### PR DESCRIPTION
**Pull request by Z-a-r-a-k-i**
…iption validate call

close #539 
close #530

### CODE REVIEW

File: go/pkg/pwapi/api_challenge-subscription-validate_test.go [hunk 1]
```diff
@@ -72,6 +73,12 @@ func TestSvc_ChallengeSubscriptionValidate(t *testing.T) {
 		assert.Equalf(t, test.input.Comment, ret.ChallengeValidation.AuthorComment, test.name)
 		assert.Equalf(t, test.expectedPassphraseIndices, ret.ChallengeValidation.Passphrases, test.name)
 		assert.NotEmptyf(t, ret.ChallengeValidation.ChallengeSubscription.Validations, test.name)
+		assert.NotNilf(t, ret.ChallengeSubscription.ClosedAt, test.name)
+		assert.Equalf(t, session.User.ID, ret.ChallengeSubscription.CloserID, test.name)
+		assert.Equalf(t, pwdb.ChallengeSubscription_Closed, ret.ChallengeSubscription.Status, test.name)
+		assert.Equalf(t, activeTeam.ID, ret.ChallengeSubscription.Team.ID, test.name)
+		assert.Equalf(t, test.input.ChallengeSubscriptionID, ret.ChallengeSubscription.ID, test.name)
+		// assert.NotEmptyf(t, ret.ChallengeSubscription.Validations, test.name)
```
Comment (Z-a-r-a-k-i) [inline go/pkg/pwapi/api_challenge-subscription-validate_test.go:81]: I cannot figure out why this test was working before.
I don't see anywhere the Validations array from ChallengeSubscription was populated previously..
Moreover repeated field in proto message aren't handled by gorm so far I know.
I'm a bit lost on this one...

Comment (moul) [reply go/pkg/pwapi/api_challenge-subscription-validate_test.go:81]: > I cannot figure out why this test was working before.

I don't understand this phrase, are you talking about the lines you just added?

Comment (moul) [reply go/pkg/pwapi/api_challenge-subscription-validate_test.go:81]: >  I don't see anywhere the Validations array from ChallengeSubscription was populated previously..


Are you talking about this: https://github.com/pathwar/pathwar/blob/master/go/pkg/pwapi/api_challenge-subscription-validate.go#L110-L120 ?

Comment (moul) [reply go/pkg/pwapi/api_challenge-subscription-validate_test.go:81]: >  Moreover repeated field in proto message aren't handled by gorm so far I know.

Yep, I confirm that gorm won't handle repeated string,int,bool well
As soon as you want to host deep objects/slices/dicts, you need to jsonify them
But from what I can see in the pwdb.proto, we only have repeated fields on relationships which is okay -> what is your problem exactly?

Comment (moul) [reply go/pkg/pwapi/api_challenge-subscription-validate_test.go:81]: >  Not sure if this assert is still relevant nor how to make it work properly.

the goal of this  unit test is to be 100% sure that a correct validation will fulfill everything well, so yes, it's very important to assert that validating a challenge will have impact on the ChallengeSubscription

Comment (moul) [reply go/pkg/pwapi/api_challenge-subscription-validate_test.go:81]: >  I'm a bit lost on this one...

Give me more details here or privately on discord, I will help you

Comment (Z-a-r-a-k-i) [reply go/pkg/pwapi/api_challenge-subscription-validate_test.go:81]: > > I cannot figure out why this test was working before.
> 
> I don't understand this phrase, are you talking about the lines you just added?

I'm talking about `assert.NotEmptyf(t, ret.ChallengeSubscription.Validations, test.name)`

Comment (Z-a-r-a-k-i) [reply go/pkg/pwapi/api_challenge-subscription-validate_test.go:81]: > > I don't see anywhere the Validations array from ChallengeSubscription was populated previously..
> 
> Are you talking about this: https://github.com/pathwar/pathwar/blob/master/go/pkg/pwapi/api_challenge-subscription-validate.go#L110-L120 ?

Here with my understanding, the validation is created in database, but I don't see anywhere this value is added to the Validations slice/array/json of any ChallengeSubscription object

Comment (Z-a-r-a-k-i) [reply go/pkg/pwapi/api_challenge-subscription-validate_test.go:81]: > > Moreover repeated field in proto message aren't handled by gorm so far I know.
> 
> Yep, I confirm that gorm won't handle repeated string,int,bool well
> As soon as you want to host deep objects/slices/dicts, you need to jsonify them
> But from what I can see in the pwdb.proto, we only have repeated fields on relationships which is okay -> what is your problem exactly?

So repeated fields work for relationships but not for type like int / bool / string ? Still not sure how this does work will investigate

**Comment by Z-a-r-a-k-i** [comment]
close #539 
close #530

**Pull request merged**